python3Packages.pyopencl: fix build

Ma27 opened the pull request:
###### Motivation for this change

Fix the recently broken build by adding `pybind11` and `pkgs.pybind11`
to the build. Also set $HOME to a temporary directory during the build
to avoid "Permission denied" errors in the build script.

This also unbreaks `sasview` and `pybitmessage`.

See also https://hydra.nixos.org/build/89037506

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

FRidh commented:
Why are both needed? That seems a bit weird.

FRidh commented on pkgs/development/python-modules/pyopencl/default.nix:
```diff
@@ -11,16 +11,18 @@
+  buildInputs = [ opencl-headers ocl-icd pkgs.pybind11 ];
 
-  propagatedBuildInputs = [ numpy cffi pytools decorator appdirs six Mako ];
+  propagatedBuildInputs = [ numpy cffi pytools decorator appdirs six Mako pybind11 ];
```
The Python `pybind11` should go into `nativeBuildInputs`

FRidh replied:
No, wait, this is non-native because it will be linked to, so `buildInputs`. @dotlambda seems we have a case here where `setup_requires` does *not* map to `nativeBuildInputs`.

dotlambda replied:
But shouldn't that linking be covered by `pkgs.pybind11`? Maybe it's our pybind11 packaging that needs fixing.

FRidh replied:
The Python version of the package essentially has the same data, the headers. We have to use the Python version because it's added in `setup_requires`, but other then that, we should be able to use any of the two in `buildInputs`.

dotlambda replied:
I see. How about we just patch it out of `setup_requires`?

dotlambda replied:
@Ma27 `pybind11` should be moved to `buildInputs`, shouldn't it?

Ma27 replied:
you're right. Updated the patch accordingly.

dotlambda commented on pkgs/development/python-modules/pyopencl/default.nix:
```diff
@@ -32,6 +34,10 @@ buildPythonPackage rec {
     substituteInPlace setup.py --replace "pytest>=2" ""
   '';
 
+  preBuild = ''
```
`preCheck`?

Ma27 commented:
WIthout `pkgs.pybind11` as build input the compilation fails with the following errors, so I'm not sure if the Python version will work:

```
gcc -Wno-unused-result -Wsign-compare -fwrapv -Wall -O3 -DNDEBUG -fPIC -DPYGPU_PACKAGE=pyopencl -DPYGPU_PYOPENCL=1 -Ipybind11/include -I/nix/store/6la3fblc497pacyy0vxhnjq1l1wlb5w7-python3-3.7.2/include/python3.7m -I/build/tmp.aabYVn0RAS/.local/include/python3.7m -I/nix/store/rirgscxdd9r8kgady8ilgpv6r7yvk388-python3.7-numpy-1.16.0/lib/python3.7/site-packages/numpy/core/include -I/nix/store/rirgscxdd9r8kgady8ilgpv6r7yvk388-python3.7-numpy-1.16.0/lib/python3.7/site-packages/numpy/core/include -I/nix/store/6la3fblc497pacyy0vxhnjq1l1wlb5w7-python3-3.7.2/include/python3.7m -c src/wrap_constants.cpp -o build/temp.linux-x86_64-3.7/src/wrap_constants.o -fvisibility=hidden -DVERSION_INFO="2018.2.2" -std=c++14 -fvisibility=hidden
In file included from src/wrap_cl.hpp:86:0,
                 from src/wrap_constants.cpp:27:
src/wrap_helpers.hpp:31:10: fatal error: pybind11/pybind11.h: No such file or directory
 #include <pybind11/pybind11.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
error: command 'gcc' failed with exit status 1
builder for '/nix/store/289qbw2j1a2n2a7b4cpbfl2r4hxln6y8-python3.7-pyopencl-2018.2.2.drv' failed with exit code 1
```

If I get this right, we want to patch the `pybind11` reference away from `setup_requires` and only use `pkgs.pybind11` for compilation?

dotlambda commented:
> If I get this right, we want to patch the pybind11 reference away from setup_requires and only use pkgs.pybind11 for compilation?

Correct.

Ma27 commented:
Hmm I'm afraid that this isn't that easy: we need `pkgs.pybind11` as build input (otherwise the compilation will fail as posted above), however we *need* `pythonPackages.pybind11` as well, otherwise the build will break with the following error:

```
...
File "setup.py", line 51, in __str__
    import pybind11
ModuleNotFoundError: No module named 'pybind11'
```

Admittedly I'm not that experienced with the `pybind11` integration, so I'm not 100% sure what else to do (instead of using the current approach from this PR).

dotlambda commented:
> ```
> ...
> File "setup.py", line 51, in __str__
>     import pybind11
> ModuleNotFoundError: No module named 'pybind11'
> ```

That probably comes from https://github.com/inducer/pyopencl/blob/0faf48a10edd086c783f794a6decdf85e8dc625b/aksetup_helper.py#L894. We'd need to patch that function to return `"${pkgs.pybind11}/include"`.

Ma27 commented:
I see. If @FRidh prefers the latter solution you proposed, I'd implement this, otherwise I'd vote for merging this :)

FRidh commented:
`pybind11/__init.py:get_include` needs to be fixed because it returns the path to the headers folder of `${python3}` instead of `${python3.pkgs.pybind11}`.

Ma27 commented:
Cross-referencing #56826 (the build is broken on the new release as well).

knedlsepp commented:
See also https://github.com/pybind/pybind11/pull/1426 for a requests to fix get_include.

FRidh commented:
@Ma27 we might use that PR as a patch.

Ma27 committed:
pythonPackages.pybind11: apply patch to properly get headers directories

It seems as the `pybind11` build code returns the Python headers
directory (where the `pybind11` headers are stored as well on traditional
setups) rather than returning the dedicated prefix[1].

An exemplary fallout is the broken build of `pyopencl`[2].

[1] https://github.com/pybind/pybind11/issues/1425
[2] https://github.com/NixOS/nixpkgs/pull/56082

Ma27 committed:
python3Packages.pyopencl: fix build

Fix the recently broken build by adding `pybind11`
to the build. Also set $HOME to a temporary directory during the build
to avoid "Permission denied" errors in the build script.

This also unbreaks `sasview` and `pybitmessage`.

See also #56826
See also https://hydra.nixos.org/build/89037506

Ma27 commented:
@FRidh applied the `pybind11` patch and updated.

dotlambda commented on pkgs/development/python-modules/pybind11/default.nix:
```diff
@@ -9,6 +9,13 @@ buildPythonPackage rec {
+  patches = [
+    (fetchpatch {
+      url = https://github.com/pybind/pybind11/commit/44a40dd61e5178985cfb1150cf05e6bfcec73042.patch;
+      sha256 = "1w55ldna72gdx481j9qxj6mah106i2p7ihvsml19l2lld54ylrfy";
```
```suggestion
      sha256 = "047nzyfsihswdva96hwchnp4gj2mlbiqvmkdnhxrfi9sji8x31ka";
```

Ma27 replied:
fixed.

Ma27 commented:
@dotlambda anything else to do? :)

dotlambda commented:
cherry picked in 2d8b128fff9ee11eb853a01a3bbd97abd6b3575e and 68c73b977fb6be3ab5ff187d12bfa786f6489119
